utils/physio.py

```diff
@@ -0,0 +1,620 @@
+            np.savetxt(physio_output, bids_file, fmt='%d', delimiter='\t')
+
+            if zip_output:
+                # Zip the file
+                gzip_command = ['gzip', '-f', physio_output]
+                # exec_command(context, gzip_command)
+                # context.custom_dict['physio-dicts'][physio]['bids_tsv'] = physio_output + '.gz'
+
+
+
+            json_output = op.join(output_dir, '{}_recording-{}_physio.json'.format(matches, label))
+            json_dict['StartTime'] = (temp_tics[0] - info_dict.ACQ_START_TICS[0]) * self.tictime
+            json_dict['Columns'] = json_header
+
+            with open(json_output, 'w') as json_file:
+                json.dump(json_dict, json_file)
+
+            # EASY PEASY LIVIN GREEZY
+
+        except Exception as e:
+            raise Exception("Error generating BIDS files for {}".format(self.type)) from e
+        pass
+
+
+
+
+
+
+
+
+
+
+
+class phys_info:
+    def __init__(self, logfile):
+        self.info_dict = {}
+        self.tic_len = 2.5e-3
+        self.ACQ_START_TICS = []
+        self.NEW_VOL_TICS = []
+        self.NEW_VOL_TIMES = []
+
+        self.info_dict = pt.log2dict(logfile)
+
+        self.NEW_VOL_TICS, self.NEW_VOL_TIMES = self.get_volume_start_tics()
+
+
+    def get_volume_start_tics(self):
+        try:
+            # Let's convert the "volume" and "ticks" array to int
+            self.info_dict['VOLUME'] = np.array(self.info_dict['VOLUME']).astype(int)
+            self.info_dict['ACQ_START_TICS'] = np.array(self.info_dict['ACQ_START_TICS']).astype(int)
+
+            # Take the difference to see when it changes and extract the ticks at those changes
+            new_vol_inds = np.where(np.diff(self.info_dict['VOLUME']))[0] + 1
+            new_vol_ticks = self.info_dict['ACQ_START_TICS'][new_vol_inds]
+            self.info_dict['NEW_VOL_TICS'] = new_vol_ticks
+            self.info_dict['NEW_VOL_TIMES'] = new_vol_ticks * self.tic_len
+
+            self.ACQ_START_TICS = np.array(self.info_dict['ACQ_START_TICS']).astype(int)
+
+            return new_vol_ticks, new_vol_ticks * self.tic_len
+
+        except Exception as e:
+            raise Exception("Error getting volume start tics for {}".format(self.logfile)) from e
+
+
+    def match_volstart_to_tic(self, tics):
+        try:
+            last_tic_vol = -1
+            new_tics = []
+            new_vol_array = np.zeros(len(tics))
+
+            for i, tic in enumerate(self.NEW_VOL_TICS):
+
+                current_tic_vol = np.argmax(tics >= tic)
+
+                if current_tic_vol > last_tic_vol:
+                    new_vol_array[current_tic_vol] = 1
+                    new_tics.append(current_tic_vol)
+
+                    last_tic_vol = current_tic_vol
+
+            return new_tics, new_vol_array
+
+        except Exception as e:
+            raise Exception("Error matching volume start tics for {}".format(self.logfile)) from e
+
+
+
+
```

Comment:
Overall,
I like the integration of your QC explorations into classes directly associated with each physio object.